server/src/main/java/org/elasticsearch/cluster/metadata/OperatorErrorMetadata.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /**
+     * Helper method to construct an {@link OperatorErrorMetadata} {@link Builder}
+     * @return {@link Builder}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Builder class for {@link OperatorErrorMetadata}
+     */
+    public static class Builder {
+        private static final String ERRORS = "errors";
+        private static final String VERSION = "version";
+        private static final String ERROR_KIND = "error_kind";
+
+        private Long version;
+        private List<String> errors;
+        private ErrorKind errorKind;
+
+        /**
+         * Default constructor
+         */
+        public Builder() {
+            this.version = 0L;
+            this.errors = new ArrayList<>();
+        }
+
+        /**
+         * Set the error metadata version
+         * @param version the metadata version of the operator content that failed to apply
+         * @return {@link Builder}
+         */
+        public Builder version(Long version) {
+            this.version = version;
+            return this;
+        }
+
+        /**
+         * Set the list of errors we encountered while processing an operator state content
+         * @param errors a list of all errors
+         * @return {@link Builder}
+         */
+        public Builder errors(List<String> errors) {
+            this.errors = errors;
+            return this;
+        }
+
+        /**
+         * Set the validation error kind
+         * @param errorKind one of {@link ErrorKind}
+         * @return {@link Builder}
+         */
+        public Builder errorKind(ErrorKind errorKind) {
+            this.errorKind = errorKind;
+            return this;
+        }
+
+        /**
+         * Builds the {@link OperatorErrorMetadata}
+         * @return {@link OperatorErrorMetadata}
+         */
+        public OperatorErrorMetadata build() {
+            return new OperatorErrorMetadata(version, errorKind, Collections.unmodifiableList(errors));
+        }
+
+        /**
+         * Serializes an {@link OperatorErrorMetadata} to xContent
+         *
+         * @param metadata {@link OperatorErrorMetadata}
+         * @param builder {@link XContentBuilder}
+         */
+        public static void toXContent(OperatorErrorMetadata metadata, XContentBuilder builder) throws IOException {
+            builder.startObject();
+            builder.field(VERSION, metadata.version);
+            builder.field(ERROR_KIND, metadata.errorKind.getKindValue());
+            builder.stringListField(ERRORS, metadata.errors);
+            builder.endObject();
+        }
+
+        /**
+         * Reads an {@link OperatorErrorMetadata} from xContent
+         *
+         * @param parser {@link XContentParser}
+         * @return {@link OperatorErrorMetadata}
+         * @throws IOException
+         */
+        public static OperatorErrorMetadata fromXContent(XContentParser parser) throws IOException {
```

Comment:
Shouldn't this be on the record, not the builder?

Reply:
Sure, I'll move them in, I was following patterns used elsewhere in the metadata.

Reply:
Actually for this one is a lot more convenient to be a builder method, we are iterating over the parser elements and we fill in the pieces of data we parse inside the builder. I had made the toXContent part of the builder too since they share the static strings, but I can remove that wrapping.